engine/crates/integration-tests/tests/federation/auth/oauth2.rs

```diff
@@ -0,0 +1,283 @@
+              "extensions": {
+                "auth": [
+                  "Access token not found"
+                ]
+              }
+            }
+          ]
+        }
+        "###);
+
+        // Proper HS256 JWT, but unrelated.
+        let response: GraphqlResponse = engine
+            .execute("query { serverVersion }")
+            .header("Authorization", "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c")
+            .await;
+        insta::assert_json_snapshot!(response, @r###"
+        {
+          "data": null,
+          "errors": [
+            {
+              "message": "Unauthorized",
+              "extensions": {
+                "auth": [
+                  "Invalid access token: No valid key found to decode the access token"
+                ]
+              }
+            }
+          ]
+        }
+        "###);
+    });
+}
+
+#[test]
+fn test_tempered_jwt() {
+    runtime().block_on(async move {
+        let github_mock = MockGraphQlServer::new(FakeGithubSchema).await;
+
+        let engine = Engine::builder()
+            .with_schema("github", &github_mock)
+            .await
+            .with_supergraph_config(format!("extend schema @authv2(providers: [{OAUTH2_PROVIDER_CONFIG}])"))
+            .finish()
+            .await;
+        let provider = OrkyHydraOpenIDProvider::default();
+        let client = provider.create_client().await;
+
+        let token = client
+            .exchange_client_credentials()
+            .add_extra_param("audience", AUDIENCE)
+            .request_async(openidconnect::reqwest::async_http_client)
+            .await
+            .unwrap()
+            .access_token()
+            .secret()
+            .clone();
+
+        let token = temper_jwt(token);
+
+        let response: GraphqlResponse = engine
+            .execute("query { serverVersion }")
+            .header("Authorization", format!("Bearer {token}"))
+            .await;
+        insta::assert_json_snapshot!(response, @r###"
+        {
+          "data": null,
+          "errors": [
+            {
+              "message": "Unauthorized",
+              "extensions": {
+                "auth": [
+                  "Invalid access token: InvalidSignature"
+                ]
+              }
+            }
+          ]
+        }
+        "###);
+    });
+}
+
+pub(super) fn temper_jwt(token: String) -> String {
+    use base64::{engine::general_purpose, Engine as _};
+    let (header, rest) = token.split_once('.').unwrap();
+    let mut header = serde_json::from_slice::<HashMap<String, serde_json::Value>>(
+        &general_purpose::URL_SAFE.decode(header).unwrap(),
+    )
+    .unwrap();
+    header.insert("typ".to_string(), serde_json::Value::String("???".to_string()));
+    let header = general_purpose::URL_SAFE.encode(serde_json::to_vec(&header).unwrap());
+    format!("{}.{}", header, rest)
```

Comment:
Minor thing, but might be a better test to tamper with the body of the JWT: probably more important to catch changes there.

Would also probably not mess with a value like `typ` that we might make use of one day - this change would potentially be rejected before we even got to the signature checking stage, making this test worse.  (I realise you're putting extensions into the response so we'd notice at the moment, but I'm not sure we should be doing that in the first place).

Reply:
good point